File: scripts/python/kmeans.py
import numpy as np
import pandas as pd
import ecopy as ep
import matplotlib.pyplot as plt
from adjustText import adjust_text
from sklearn.cluster import KMeans

to_print = True

# Leitura da matriz de dados de abundância
matriz = pd.read_csv("especies.csv", header=0, index_col=0, encoding="utf8")

# Transposição da matriz de dados
matriz = matriz.transpose()

# Transformação logarítmica
matriz_log = matriz.apply(lambda x: np.log10(x + 1)) # Logaritmos decimais

# Cálculo da distância de Bray-Curtis entre as amostras
# utilizando a função distance() do pacote "ecopy"
X = ep.distance(matriz_log, method="bray")

# Número de agrupamentos 
kmeans_model = KMeans(n_clusters=9)

# Ajuste dos dados de entrada
kmeans_model.fit(X)

# Obtenção dos rótulos dos agrupamentos
labels = kmeans_model.predict(X)

# Valores dos centroides
centroids = kmeans_model.cluster_centers_

# Saída dos resultados
print(labels)
print(centroids)

# Diagrama de dispersão
fig, ax = plt.subplots()
x = list(centroids[:,0])
y = list(centroids[:,1])
if to_print:
    plt.scatter(x, y, color="black", s=20)
else:
    plt.scatter(x, y, color="blue", s=20)
texts = []
for i, txt in enumerate(list(matriz.index)):
    texts.append(ax.text(x[i], y[i], txt, size=7))
adjust_text(texts)
plt.tight_layout()
plt.show()

File: scripts/r/kmeans.R
to_print <- TRUE

# Leitura dos dados de abundância
matriz <- read.csv("especies.csv", row.names=1) 

# Transposição da matriz de dados
matriz <- t(matriz)

# Transformação logarítmica da matriz de abundância
matriz_log <- log10(matriz + 1)

# Cálculo da distância de Bray-Curtis entre as amostras
# utilizando a função "vegdist" do pacote "vegan"
library(vegan)
dist <- vegdist(matriz_log, method="bray") 

# Análise de agrupamentos divisiva
# usando a função "means" do pacote "stats"
cl <- kmeans(dist, centers=2, nstart=8)

# Saída dos resultados
print(cl)

# Diagrama de dispersão
if (to_print) {
  plot(cl$centers[1,], cl$centers[2,], pch=18:19, xlab="Distancia", ylab="Distancia", 
       main=paste("Compactness = ", round(cl$betweenss/cl$totss*100.0, digits=1), "%"))
  text(cl$centers[1,], cl$centers[2,], labels=rownames(matriz), cex=0.5, pos=4)
  } else {
    plot(cl$centers[1,], cl$centers[2,], col=cl$cluster, pch=19, xlab="Distancia", ylab="Distancia", 
         main=paste("Compactness = ", round(cl$betweenss/cl$totss*100.0, digits=1), "%"))
  text(cl$centers[1,], cl$centers[2,], labels=rownames(matriz), cex=0.5, pos=3)
}

# Análise de agrupamentos não-hierárquica
# usando a função "pam" do pacote "cluster"
library(cluster)
pamx <- pam(dist, 2, diss=TRUE)
print(summary(pamx))
plot(pamx, main="")
kmcol <- pamx$clustering